src/session/key.ts

```diff
@@ -0,0 +1,123 @@
+
+export const CREATE_DATA_SET_TYPEHASH = EIP712_TYPE_HASHES.CreateDataSet
+export const ADD_PIECES_TYPEHASH = EIP712_TYPE_HASHES.AddPieces
+export const SCHEDULE_PIECE_REMOVALS_TYPEHASH = EIP712_TYPE_HASHES.SchedulePieceRemovals
+export const DELETE_DATA_SET_TYPEHASH = EIP712_TYPE_HASHES.DeleteDataSet
+
+export const PDP_PERMISSIONS = [
+  CREATE_DATA_SET_TYPEHASH,
+  ADD_PIECES_TYPEHASH,
+  SCHEDULE_PIECE_REMOVALS_TYPEHASH,
+  DELETE_DATA_SET_TYPEHASH,
+]
+
+export const PDP_PERMISSION_NAMES: Record<string, string> = {
+  [CREATE_DATA_SET_TYPEHASH]: 'CreateDataSet',
+  [ADD_PIECES_TYPEHASH]: 'AddPieces',
+  [SCHEDULE_PIECE_REMOVALS_TYPEHASH]: 'SchedulePieceRemovals',
+  [DELETE_DATA_SET_TYPEHASH]: 'DeleteDataSet',
+}
+
+export class SessionKey {
+  private readonly _provider: ethers.Provider
+  private readonly _registry: ethers.Contract
+  private readonly _signer: ethers.Signer
+  private readonly _owner: ethers.Signer
+
+  public constructor(
+    provider: ethers.Provider,
+    sessionKeyRegistryAddress: string,
+    signer: ethers.Signer,
+    owner: ethers.Signer
+  ) {
+    this._provider = provider
+    this._registry = new ethers.Contract(sessionKeyRegistryAddress, CONTRACT_ABIS.SESSION_KEY_REGISTRY, owner)
+    this._signer = signer
+    this._owner = owner
+  }
+
+  getSigner(): ethers.Signer {
+    return this._signer
+  }
+
+  /**
+   * Queries current permission expiries from the registry
+   */
+  async fetchExpiries(permissions: string[] = PDP_PERMISSIONS): Promise<Record<string, bigint>> {
+    const network = await getFilecoinNetworkType(this._provider)
+
+    const multicall = new ethers.Contract(
+      CONTRACT_ADDRESSES.MULTICALL3[network],
+      CONTRACT_ABIS.MULTICALL3,
+      this._provider
+    )
+    const registryInterface = new ethers.Interface(CONTRACT_ABIS.SESSION_KEY_REGISTRY)
+
+    const [ownerAddress, signerAddress, registryAddress] = await Promise.all([
+      this._owner.getAddress(),
+      this._signer.getAddress(),
+      this._registry.getAddress(),
+    ])
+
+    // Prepare multicall batch
+    const calls: Array<{ target: string; allowFailure: boolean; callData: string }> = []
+    for (const permission of permissions) {
+      calls.push({
+        target: registryAddress,
+        allowFailure: true,
+        callData: registryInterface.encodeFunctionData('authorizationExpiry', [
+          ownerAddress,
+          signerAddress,
+          permission,
+        ]),
+      })
+    }
+
+    // Execute multicall
+    const results = await multicall.aggregate3.staticCall(calls)
+
+    const expiries: Record<string, bigint> = {}
+    for (let i = 0; i < permissions.length; i++) {
+      expiries[PDP_PERMISSIONS[i]] = registryInterface.decodeFunctionResult(
+        'authorizationExpiry',
+        results[i].returnData
+      )[0]
+    }
+    return expiries
+  }
+
+  /**
+   * Owner authorizes signer with permissions until expiry
```

Comment:
```suggestion
   * Authorize signer with permissions until expiry. This can also be used to
   * renew or invalidate existing authorization by updating the expiry.
   *
```

Reply:
For invalidate we can make a wrapper for `invalidate()`.